diff --git a/mongoose/Makefile b/mongoose/Makefile
index 27ebd812..4fae19c3 100644
--- a/mongoose/Makefile
+++ b/mongoose/Makefile
@@ -11,6 +11,7 @@ KBUILD_CFLAGS += -DMG_ENABLE_LOG=0
 KBUILD_CFLAGS += -DMG_ENABLE_MD5=1
 KBUILD_CFLAGS += -DMG_ENABLE_THREADS=1
 KBUILD_CFLAGS += -DMG_MAX_RECV_BUF_SIZE=262144
+KBUILD_CFLAGS += -DMG_ENABLE_HTTP_CGI=1
 ifneq ($(CONFIG_MONGOOSEIPV6),)
 KBUILD_CFLAGS += -DMG_ENABLE_IPV6=1
 endif
